Keep track of connection type in conntracker.go

leeavital opened the pull request:
### What does this PR do?

Include L4 proto in the connection key used by conntracker.go

### Motivation

the old key for each conntrack entry was `(sourceIP, sourcePort)`.  Because it's possible to use have UDP and TCP  connections have the same sourceIP/sourcePort, a bunch of TCP connections were incorrectly . attached to UDP translations.

Also includes an integration test which uses IP tables to simulate a DNAT-ed connection. The integration test needs to be run as root, and will not run otherwise.

### Additional Notes

Anything else we should know when reviewing?

leeavital committed:
Keep track of connection type in conntracker.go

And add an integration test.

leeavital committed:
Add more tests

leeavital committed:
Fix ineffassign in conntracker_test.go

leeavital committed:
make tests only run with ebpf tag

sunhay commented on pkg/ebpf/netlink/conntracker.go:
```diff
@@ -159,13 +161,13 @@ func newConntrackerOnce(procRoot string, deleteBufferSize, maxStateSize int) (Co
 }
 
-func (ctr *realConntracker) GetTranslationForConn(ip util.Address, port uint16) *IPTranslation {
+func (ctr *realConntracker) GetTranslationForConn(ip util.Address, port uint16, proto uint8) *IPTranslation {
```
Any reason we can't just use `ConnectionType` here instead of uint8?

leeavital replied:
I had originally tried to use `ebpf.ConnectionType` but it caused a cyclic dependency between `ebpf` and `ebpf/netlink`.  


We could use the `process.ConnectionType` but I would still need to convert between `connection` and `process` types.  Now that I'm tying it out, I think that's a little easier to follow. I'll do that.  


Long term, I think we should try getting rid of `ebpf.ConnectionType` together in favor of `process.ConnectionType` but that's  a bigger code change.

sunhay commented on pkg/ebpf/netlink/conntracker.go:
```diff
@@ -401,5 +403,13 @@ func formatKey(c ct.Conn) (k connKey) {
 	if port, err := c.Uint16(ct.AttrOrigPortSrc); err == nil {
 		k.port = NtohsU16(port)
 	}
+	if proto, err := c.Uint8(ct.AttrOrigL4Proto); err == nil {
```
Looks like there's a number of other options here - can we short circuit the entire entry if it's not UDP/TCP where this is called? It looks like we're storing 0 ortherwise for `k.proto`

sunhay commented on pkg/ebpf/netlink/conntracker.go:
```diff
@@ -38,6 +37,9 @@ type Conntracker interface {
 	port uint16
+
+	// the protocol of the connection, using the same values as specified in the agent payload.
+	proto uint8
```
nit: lets call this `transport` like we do in the backend. In the future we will likely be using `protocol` for deeper inspection like `http`, etc.

leeavital committed:
Add some error checking to formatKey

leeavital committed:
Address review feedback:

- be more type safe (use process.ConnectionType instead of uint8)
- error checking for bad protocols

leeavital committed:
Make tests more type safe

leeavital committed:
Fix tracer.go
